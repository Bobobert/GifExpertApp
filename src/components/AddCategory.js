import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) => {
        e.preventDefault(); // avoids to refresh the page

        if (inputValue.trim().length > 1) {
            setCategories( categories => [inputValue, ...categories]);
            setInputValue('');
        }

    }

    return <form onSubmit={ handleSubmit }>
        <input 
            type="text"
            value={ inputValue }
            onChange={ handleInputChange }
        />
        </form>
}
//element from agroups every input element inside, so it works
// the same as a fragment in this snippet

AddCategory.propTypes = {
    setCategories : PropTypes.func.isRequired,
}