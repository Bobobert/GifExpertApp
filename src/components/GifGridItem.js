import React from 'react';

export const GifGridItem = ({id, title, url}) => {

    const handleCopy = () => {
        navigator.clipboard.writeText(`https://giphy.com/embed/${id}`).then(
        console.log(`Copied to clipboard ${title} !`));
    }

    return (
        <div className="card animate__animated animate__fadeInDown animate__fast" onClick={handleCopy}>
            <img src={ url } alt={ title }/>
            <div className='overlay'>
                <div className='text'>
                    This GIF!<br/>{title}
                </div>
            </div>

        </div>
    )
}
// extracted more info and style from this https://www.w3schools.com/css/css3_images.asp
