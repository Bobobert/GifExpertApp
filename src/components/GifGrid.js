import React from 'react'
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    //const [images, setImages] = useState([]);
    const { data:images, loading } = useFetchGifs( category );

    // useEffect(() => {
    //     getGifs();
    // }, []); // este snip sirve para bloquear la ejecucion de funciones
    // cuando son ejecutadas/ renderiadas por primera vez, al menos.
    // El problema estÃ¡ en que todo  este componente se ejecuta cuando se realiza
    // un cambio.

    // useEffect( () => {
    //     getGifs( category ).then( setImages );
    // }, [ category ]); // run just once for []
    // // or when category changes [ category ]
    
    return <>
            <h3> { category } </h3>
            { loading && <p className='animate__animated animate__shakeY'>Cargando . . .</p> }
            <div className='card-grid'>
                {images.map( img => 
                    <GifGridItem key={img.id}{...img} />
                    )}
            </div>
        </>
}
