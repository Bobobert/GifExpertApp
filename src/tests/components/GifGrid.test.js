import { render, screen } from "@testing-library/react"
import { GifGrid } from "../../components/GifGrid"
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe('Pruebas para el <GifGrid />', () => {

    const categoryTest = 'not-a-single-gif-in-this-dasdas';

    test('debe tener un comportamiento nominal(matchSnapshot)', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });

        const {container} = render(<GifGrid category={categoryTest} />);
        //screen.debug();
        //console.log( screen.queryByText(/cargando/i), 'erste' );
        expect( container ).toMatchSnapshot();

    })

    test('debe mostrar items cuando se cargan imagenes de useFetchGifs', () => {

        const gifs = [{
            id: 'TEST-ID0',
            url: 'https://localhost/any/any.jpg',
            title: 'Any Test Title',
        },
        {
            id: 'TEST-ID1',
            url: 'https://localhost/any/anything.jpg',
            title: 'Any Test Title but larger',
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });

        const {container} = render(<GifGrid category={categoryTest} />);
        //expect( container ).toMatchSnapshot();
        expect( screen.queryByText(/cargando/i) ).toBeNull();
        screen.findAllByTestId('GifGridItem').then( recieved => {
            expect( recieved.length ).toBe( gifs.length )
        });


    })

})