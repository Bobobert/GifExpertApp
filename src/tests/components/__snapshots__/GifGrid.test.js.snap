// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para el <GifGrid /> debe tener un comportamiento nominal(matchSnapshot) 1`] = `
<div>
  <h3>
     
    not-a-single-gif-in-this-dasdas
     
  </h3>
  <p
    class="animate__animated animate__shakeY"
  >
    Cargando . . .
  </p>
  <div
    class="card-grid"
    data-testid="GifGrid"
  />
</div>
`;
